src/coap_client.c

```diff
@@ -26,6 +26,46 @@ bool golioth_client_is_connected(struct golioth_client *client)
     return client->session_connected;
 }
 
+void golioth_coap_next_token(uint8_t token[GOLIOTH_COAP_TOKEN_LEN])
+{
+    static golioth_sys_mutex_t token_mut;
+    static uint8_t stored_token[GOLIOTH_COAP_TOKEN_LEN] = {0};
+
+    if (!token_mut)
+    {
+        token_mut = golioth_sys_mutex_create();
+
+        uint32_t rnd = 0;
+        golioth_sys_srand(golioth_sys_now_ms());
+
+        for (int i = 0; i < GOLIOTH_COAP_TOKEN_LEN; i++)
+        {
+            if (i % 4 == 0)
+            {
+                rnd = golioth_sys_rand();
+            }
+
+            stored_token[i] = (uint8_t) (rnd >> (i % 4));
+        }
+    }
+
+    golioth_sys_mutex_lock(token_mut, GOLIOTH_SYS_WAIT_FOREVER);
+
+    for (int i = 0; i < GOLIOTH_COAP_TOKEN_LEN; i++)
+    {
+        stored_token[i]++;
+
+        if (stored_token[i] != 0)
+        {
+            break;
+        }
+    }
```

Comment:
This was designed to allow for a token of variable length (determined at runtime). If we move to using uint64_t, that variable configuration needs to be removed. Is that your preference?

Reply:
I'm ok with fixing the token size, but I also think it would still be clearer to use a uint64 and mod or mask the value to keep it to the selected size.

Reply:
I've worked on changing the storage of a token to uint64_t [in another draft PR](https://github.com/golioth/golioth-firmware-sdk/pull/697). However, it feels pretty awkward and there are a couple of things on my mind:

1. libcoap regulary returns tokens (generated and from responses) that are smaller than 8 bytes (I've seen 4 byte and 2 byte as well). Getting rid of the array and array length breaks the port and needed workarounds when comparing tokens. I haven't made the same workarounds for Zephyr CoAP but now I worry I need to.
2. When casting an 8-byte array to a uint64_t, does the endianness of the system matter? I would think that in little-endian, if you cast a uint64_t to (uint8_t *) you'll end up with the even and odd bytes flipped in the array.

Reply:
Okay, as we discussed out-of-band, I updated this to copy the array to uint64_t, increment, then copy back.